### PR TITLE
add plaid support to bank account creation

### PR DESCRIPTION
**Pull request by mgoren**

### CODE REVIEW

File: app/assets/javascripts/bank_accounts.js.erb [hunk 1]
```diff
@@ -1,4 +1,42 @@
...
+          method: "POST",
+          url: "<%= url_for(bank_accounts_path) %>",
+          data: { public_token: public_token, account_id: metadata.account_id }
+        });
+      },
+      onExit: function(err, metadata) {
+        $("#bank-account-options").show()
+        $("#bank-account-please-wait").hide();
+        if (err != null) {
+          alert("Plaid API error");
```
Comment (michaelrkn) [inline app/assets/javascripts/bank_accounts.js.erb]: Maybe display the contents of `err`?

File: app/controllers/bank_accounts_controller.rb [hunk 2]
```diff
@@ -6,9 +6,21 @@ def new
   end
 
   def create
-    @bank_account = BankAccount.new(bank_account_params.merge(student: current_student))
-    unless @bank_account.save
-      render :new
+    if params[:public_token] && params[:account_id] # Creating bank account through Plaid
```
Comment (michaelrkn) [inline app/controllers/bank_accounts_controller.rb]:
```
if verified_with_plaid?
...

private

def verified_with_plaid?
  params[:public_token] && params[:account_id]
end
```

File: app/controllers/bank_accounts_controller.rb [hunk 3]
```diff
@@ -6,9 +6,21 @@ def new
...
+      stripe_token = BankAccount.exchange_plaid_token(params[:public_token], params[:account_id])
```
Comment (michaelrkn) [inline app/controllers/bank_accounts_controller.rb]: can you just pass these params into `BankAccount.new` and let it do this work?

File: app/controllers/bank_accounts_controller.rb [hunk 4]
```diff
@@ -6,9 +6,21 @@ def new
...
+        flash[:notice] = "Your bank account has been confirmed."
+      else
+        flash[:alert] = "There was a problem linking your bank account."
+      end
+      render js: "window.location.pathname ='#{payment_methods_path}'"
+    else # Creating bank account manually
+      @bank_account = BankAccount.new(bank_account_params.merge(student: current_student))
+      unless @bank_account.save
+        render :new
+      end
```
Comment (michaelrkn) [inline app/controllers/bank_accounts_controller.rb]: I'd like to see most of this logic get pushed down to the model. regardless of whether its plaid or manual, just pass in the params from the request, let the model sort out what to do, and then redirect appropriately based on what type of verification happens next.

File: app/controllers/bank_accounts_controller.rb [hunk 5]
```diff
@@ -6,9 +6,26 @@ def new
...
-    unless @bank_account.save
-      render :new
+    # bank_account_params = {} if bank_account_params.nil?
+    # binding.pry
+    @bank_account = BankAccount.new(bank_account_params.merge(student: current_student, plaid_public_token: params[:public_token], plaid_account_id: params[:account_id]))
+    if @bank_account.save
+      respond_to do |format|
+        format.html { render :create } # setup manually
+        format.js do # setup via Plaid
+          @bank_account.ensure_primary_method_exists
```
Comment (michaelrkn) [inline app/controllers/bank_accounts_controller.rb]: why isn't this in the model?

File: app/assets/javascripts/bank_accounts.js.erb [hunk 6]
```diff
@@ -1,4 +1,43 @@
...
+      clientName: 'Epicodus',
+      product: ['auth'],
+      env: '<%= Rails.env.production? ? 'production' : 'sandbox' %>',
+      key: '<%= ENV['PLAID_PUBLIC_KEY'] %>',
+      onSuccess: function(public_token, metadata) {
+        $.ajax({
+          method: "POST",
+          dataType: "script",
+          url: "<%= url_for(bank_accounts_path) %>",
+          data: { public_token: public_token, account_id: metadata.account_id }
```
Comment (michaelrkn) [inline app/assets/javascripts/bank_accounts.js.erb]: change this to:

```
data: { bank_account: { plaid_public_token: public_token, plaid_account_id: metadata.account_id } }
```

File: app/controllers/bank_accounts_controller.rb [hunk 7]
```diff
@@ -6,9 +6,26 @@ def new
   end
 
   def create
-    @bank_account = BankAccount.new(bank_account_params.merge(student: current_student))
-    unless @bank_account.save
-      render :new
+    # bank_account_params = {} if bank_account_params.nil?
+    # binding.pry
+    @bank_account = BankAccount.new(bank_account_params.merge(student: current_student, plaid_public_token: params[:public_token], plaid_account_id: params[:account_id]))
```
Comment (michaelrkn) [inline app/controllers/bank_accounts_controller.rb]: change this to:

```
@bank_account = BankAccount.new(bank_account_params.merge(student: current_student)
```

File: app/controllers/bank_accounts_controller.rb [hunk 8]
```diff
@@ -29,4 +46,8 @@ def update
   def bank_account_params
     params.require(:bank_account).permit(:first_deposit, :second_deposit, :stripe_token)
```
Comment (michaelrkn) [inline app/controllers/bank_accounts_controller.rb]: change this to:

```
params.require(:bank_account).permit(:first_deposit, :second_deposit, :stripe_token, :plaid_account_id, : plaid_public_token)
```

File: app/assets/javascripts/bank_accounts.js.erb (hunk 6 above)
Comment (mgoren) [reply app/assets/javascripts/bank_accounts.js.erb]: Got it working by sticking that inside bank_account hash, as follows. :)
data: { bank_account: { plaid_public_token: public_token, plaid_account_id: metadata.account_id } }

File: app/controllers/bank_accounts_controller.rb [hunk 9]
```diff
@@ -27,6 +41,10 @@ def update
 
 private
   def bank_account_params
-    params.require(:bank_account).permit(:first_deposit, :second_deposit, :stripe_token)
+    params.require(:bank_account).permit(:first_deposit, :second_deposit, :stripe_token, :plaid_account_id, :plaid_public_token)
+  end
+
+  def verified_with_plaid?
```
Comment (michaelrkn) [inline app/controllers/bank_accounts_controller.rb]: This is no longer used.

File: app/models/bank_account.rb [hunk 10]
```diff
@@ -1,7 +1,13 @@
 class BankAccount < PaymentMethod
   before_update :verify_account, unless: ->(bank_account) { bank_account.verified? }
 
-  attr_accessor :first_deposit, :second_deposit
+  attr_accessor :first_deposit, :second_deposit, :plaid_public_token, :plaid_account_id
+
+  def initialize(params = {})
+    @plaid_public_token = params[:plaid_public_token]
+    @plaid_account_id = params[:plaid_account_id]
```
Comment (michaelrkn) [inline app/models/bank_account.rb]: are these instance variables actually used anywhere? it seems like not.

Comment (mgoren) [reply app/models/bank_account.rb]: Oh cool! I hadn't realized that the params merge works with just the attr_accessors (for non-db-backed variables). I've now eliminated the initialize method entirely, so much cleaner. :)

**Pull request closed without merging**